src/newTest/java/org/opensearch/test/framework/cluster/LocalOpenSearchCluster.java

```diff
@@ -0,0 +1,610 @@
+        List<CompletableFuture<String>> futures = new ArrayList<>();
+
+        for (NodeSettings nodeSettings : nodeSettingList) {
+            Node node = new Node(nodeSettings, transportPortIterator.next(), httpPortIterator.next());
+            futures.add(node.start());
+        }
+
+        return CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
+    }
+
+    public ClusterInfo waitForCluster(ClusterHealthStatus status, TimeValue timeout, int expectedNodeCount) throws IOException {
+        ClusterInfo clusterInfo = new ClusterInfo();
+        Client client = clientNode().getInternalNodeClient();
+
+        try {
+            log.debug("waiting for cluster state {} and {} nodes", status.name(), expectedNodeCount);
+            AdminClient adminClient = client.admin();
+
+            final ClusterHealthResponse healthResponse = adminClient.cluster().prepareHealth().setWaitForStatus(status).setTimeout(timeout)
+                    .setMasterNodeTimeout(timeout).setWaitForNodes("" + expectedNodeCount).execute().actionGet();
+
+            if (log.isDebugEnabled()) {
+                log.debug("Current ClusterState:\n{}", Strings.toString(healthResponse));
+            }
+
+            if (healthResponse.isTimedOut()) {
+                throw new IOException(
+                        "cluster state is " + healthResponse.getStatus().name() + " with " + healthResponse.getNumberOfNodes() + " nodes");
+            } else {
+                log.debug("... cluster state ok {} with {} nodes", healthResponse.getStatus().name(), healthResponse.getNumberOfNodes());
+            }
+
+            assertEquals(expectedNodeCount, healthResponse.getNumberOfNodes());
+
+            final NodesInfoResponse nodesInfoResponse = adminClient.cluster().nodesInfo(new NodesInfoRequest()).actionGet();
+
+            final List<NodeInfo> nodes = nodesInfoResponse.getNodes();
+
+            final List<NodeInfo> masterNodes = nodes.stream()
+                    .filter(nodeInfo -> nodeInfo.getNode().getRoles().contains(DiscoveryNodeRole.CLUSTER_MANAGER_ROLE)).collect(Collectors.toList());
+
+            final List<NodeInfo> dataNodes = nodes.stream().filter(nodeInfo -> nodeInfo.getNode().getRoles().contains(DiscoveryNodeRole.DATA_ROLE)
+                    && !nodeInfo.getNode().getRoles().contains(DiscoveryNodeRole.CLUSTER_MANAGER_ROLE)).collect(Collectors.toList());
+
+            final List<NodeInfo> clientNodes = nodes.stream()
+                    .filter(nodeInfo -> !nodeInfo.getNode().getRoles().contains(DiscoveryNodeRole.CLUSTER_MANAGER_ROLE)
+                            && !nodeInfo.getNode().getRoles().contains(DiscoveryNodeRole.DATA_ROLE))
+                    .collect(Collectors.toList());
+
+            for (NodeInfo nodeInfo : masterNodes) {
+                final TransportAddress is = nodeInfo.getInfo(TransportInfo.class).getAddress().publishAddress();
+                clusterInfo.nodePort = is.getPort();
```

Comment:
This value is getting overwritten all over the place, seems like there could be a cleaner way to get this from the list of nodes?